### ISSUE TITLE
Multiple problems with the plugin

### ISSUE THREAD

**Issue by toma-soft** [kreuzwerker/terraform-provider-docker#138]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and docker Provider) Version
0.13.4, 0.14.8

### Affected Resource(s)

2.10.0, 2.11.0

* `Docker version 19.03.12, build 48a66213fe`

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    docker = {
      source = "kreuzwerker/docker"
      version = "2.11.0"
    }
  }
}
provider "docker" {
  host = "npipe:////.//pipe//docker_engine"

  registry_auth {
    # some settings
  }
}

data "docker_registry_image" "jenkins-image" {
  name = "some-registry/some-repository/jenkins:version"
}

resource "docker_image" "jenkins" {
  name = data.docker_registry_image.jenkins-image.name
  pull_triggers = [data.docker_registry_image.jenkins-image.sha256_digest]
  keep_locally = true
}  

resource "docker_volume" "prod-jenkins" {
  name = "prod-jenkins"
}

resource "docker_service" "prod-jenkins" {
  name = "prod-jenkins"

  task_spec {
    container_spec {
      image = data.docker_registry_image.jenkins-image.name

      mounts {
        target    = "/var/jenkins_home"
        source    = "prod-jenkins"
        type      = "volume"
        read_only = false
      }
      env = {
        # some env variables
      }
    }
  }
  endpoint_spec {
    ports {
      # some settings
    }
    ports {
      # some settings
    }
  }
}
```

### Debug Output
No special warnings or errors in debug logs.

### Panic Output
N/A

### Expected Behaviour
All resources are created, removed or updated.

### Actual Behaviour

Many different issues:

`#1 - timeout` 
```
D:\some\path>terraform apply -auto-approve -no-color 
data.docker_registry_image.jenkins-image: Refreshing state...

Error: timeout while waiting for plugin to start


[Pipeline] echo
Skipping terraform-jenkins-repo\production-jenkins2...
[Pipeline] echo
script returned exit code 1
```

`#2 - refreshing was not responding`

```
D:\some\path>terraform plan -no-color 
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.docker_registry_image.jenkins-image: Refreshing state... [id=sha256:7892bfdb1d0114fbfd7b12a5d9c298c433cbdd24316f7130b465d76a70ed5b3d]
docker_volume.dev-jenkins2: Refreshing state... [id=dev-jenkins2]
docker_image.jenkins: Refreshing state... [id=sha256:a518becb9e71f68686eff2c90f729f28fe561d44da77842aa09d4bcccb8ba537some-registry/jenkins:2.263.4-release14]
docker_service.dev-jenkins2: Refreshing state... [id=o7sszeydr2sogox686vrp6fhy]
Aborted by admin
Sending interrupt signal to process
Interrupt received.
Please wait for Terraform to exit or data loss may occur.
Gracefully shutting down...
script returned exit code -1
```
`#3 - issue with loading`

```
D:\some\path>terraform plan -no-color 

Error: Could not load plugin


Plugin reinitialization required. Please run "terraform init".

Plugins are external binaries that Terraform uses to access and manipulate
resources. The configuration provided requires plugins which can't be located,
don't satisfy the version constraints, or are otherwise incompatible.

Terraform automatically discovers provider requirements from your
configuration, including providers used in child modules. To see the
requirements and constraints, run "terraform providers".

Failed to instantiate provider "registry.terraform.io/kreuzwerker/docker" to
obtain schema: timeout while waiting for plugin to start
```
`#4 - TLS timeout`

```
Initializing the backend...Initializing provider plugins...
- Finding kreuzwerker/docker versions matching "2.11.0"...
- Installing kreuzwerker/docker v2.11.0...
Error: Failed to install providerError while installing kreuzwerker/docker v2.11.0: Get
"https://github.com/kreuzwerker/terraform-provider-docker/releases/download/v2.11.0/terraform-provider-docker_2.11.0_windows_amd64.zip":
net/http: TLS handshake timeout
```
`#5 - removing docker_volume`
```
Error: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
```

### Steps to Reproduce

1. Just use `terraform apply` or `terraform plan` - the issues above seem to be non-deterministic, sometimes I have to re-start the pipeline several times, and sometimes some builds in a row are successful.

### References

* #0000

**Comment by suzuki-shunsuke** [comment]
Regarding to `#3 - issue with loading`, you should run `terraform init`. Please read the error message.

```
D:\some\path>terraform plan -no-color 

Error: Could not load plugin


Plugin reinitialization required. Please run "terraform init".
```

**Comment by suzuki-shunsuke** [comment]
> Debug Output
> No special warnings or errors in debug logs.

Please share us debug output.
Otherwise, it is difficult to investigate.

**Comment by toma-soft** [comment]
> Regarding to `#3 - issue with loading`, you should run `terraform init`. Please read the error message.
> 
> ```
> D:\some\path>terraform plan -no-color 
> 
> Error: Could not load plugin
> 
> 
> Plugin reinitialization required. Please run "terraform init".
> ```

Yes, I haven't mention that `terraform init` is executed. Here is how I implemented it:

```
bat "terraform init ${terraformParams}"
bat "terraform plan ${terraformParams}"

try{
  bat "terraform apply -auto-approve ${terraformParams}"      
}catch(Exception e){
  echo "Skipping $pathToModule..."
  echo e.message
  currentBuild.result = 'UNSTABLE'
}
```

**Comment by toma-soft** [comment]
> > Debug Output
> > No special warnings or errors in debug logs.
> 
> Please share the debug output with us.
> Otherwise, it is difficult to investigate.

Sure, I will post here debug logs as soon as I get the error message.

**Comment by toma-soft** [comment]
Debug logs from the very first run:

```
2021/03/11 13:22:12 [WARN] Provider "registry.terraform.io/kreuzwerker/docker" produced an invalid plan for docker_service.dev-jenkins, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .mode: block count in plan (1) disagrees with count in config (0)
      - .task_spec[0].container_spec[0].dns_config: block count in plan (1) disagrees with count in config (0)
      - .task_spec[0].container_spec[0].healthcheck: block count in plan (1) disagrees with count in config (0)
      - .task_spec[0].placement: block count in plan (1) disagrees with count in config (0)
      - .task_spec[0].resources: block count in plan (1) disagrees with count in config (0)
2021-03-11T13:22:12.702+0100 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe pid=18264
2021-03-11T13:22:12.702+0100 [DEBUG] plugin: plugin exited
2021/03/11 13:22:12 [INFO] backend/local: apply calling Apply
2021/03/11 13:22:12 [INFO] terraform: building graph: GraphTypeApply
2021/03/11 13:22:12 [DEBUG] ProviderTransformer: "docker_image.jenkins-aec (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/11 13:22:12 [DEBUG] ProviderTransformer: "docker_volume.dev-jenkins (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/11 13:22:12 [DEBUG] ProviderTransformer: "data.docker_registry_image.jenkins-aec-image (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/11 13:22:12 [DEBUG] ProviderTransformer: "docker_service.dev-jenkins (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/11 13:22:12 [DEBUG] ReferenceTransformer: "docker_service.dev-jenkins (expand)" references: []
2021/03/11 13:22:12 [DEBUG] ReferenceTransformer: "docker_image.jenkins-aec (expand)" references: []
2021/03/11 13:22:12 [DEBUG] ReferenceTransformer: "docker_volume.dev-jenkins (expand)" references: []
2021/03/11 13:22:12 [DEBUG] ReferenceTransformer: "data.docker_registry_image.jenkins-aec-image (expand)" references: []
2021/03/11 13:22:12 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/kreuzwerker/docker\"]" references: []
2021/03/11 13:22:12 [DEBUG] pruneUnusedNodes: docker_image.jenkins-aec (expand) is no longer needed, removing
2021/03/11 13:22:12 [DEBUG] pruneUnusedNodes: docker_volume.dev-jenkins (expand) is no longer needed, removing
2021/03/11 13:22:12 [DEBUG] pruneUnusedNodes: data.docker_registry_image.jenkins-aec-image (expand) is no longer needed, removing
2021/03/11 13:22:12 [DEBUG] pruneUnusedNodes: docker_service.dev-jenkins (expand) is no longer needed, removing
2021/03/11 13:22:12 [DEBUG] Starting graph walk: walkApply
2021-03-11T13:22:12.705+0100 [INFO]  plugin: configuring client automatic mTLS
2021-03-11T13:22:12.740+0100 [DEBUG] plugin: starting plugin: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe args=[.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe]
2021-03-11T13:22:12.872+0100 [DEBUG] plugin: plugin started: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe pid=11624
2021-03-11T13:22:12.872+0100 [DEBUG] plugin: waiting for RPC address: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe
2021-03-11T13:22:13.068+0100 [INFO]  plugin.terraform-provider-docker_v2.10.0.exe: configuring server automatic mTLS: timestamp=2021-03-11T13:22:13.061+0100
2021/03/11 13:23:12 [ERROR] eval: *terraform.EvalInitProvider, err: timeout while waiting for plugin to start
2021/03/11 13:23:12 [ERROR] eval: *terraform.EvalSequence, err: timeout while waiting for plugin to start
2021-03-11T13:23:12.887+0100 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.10.0/windows_amd64/terraform-provider-docker_v2.10.0.exe pid=11624 error="exit status 1"

Error: timeout while waiting for plugin to start


2021-03-11T13:23:12.938+0100 [WARN]  plugin: plugin failed to exit gracefully
```

**Comment by suzuki-shunsuke** [comment]
Thank you for sharing debug log.

Maybe this issue is related.
https://github.com/hashicorp/terraform/issues/25133

**Comment by toma-soft** [comment]
Is it about the Go installation on Windows ? The mentioned thread is still open so it looks like there no "official" answer.

**Comment by suzuki-shunsuke** [comment]
This provider is cross-compiled on Linux, so I think Go version on Windows has nothing to do with this problem.

Do you have any firewall software on your Windows system that might be blocking requests like these?
ref. https://github.com/hashicorp/terraform/issues/25133#issuecomment-639131813

**Comment by suzuki-shunsuke** [comment]
https://github.com/hashicorp/terraform/issues/25133#issuecomment-796757686

**Comment by toma-soft** [comment]
> This provider is cross-compiled on Linux, so I think Go version on Windows has nothing to do with this problem.
> https://github.com/kreuzwerker/terraform-provider-docker/blob/97ed39e2d99c12240a57898feb4cc9c2adce685c/.github/workflows/compile.yaml#L15-L31
> 
> Do you have any firewall software on your Windows system that might be blocking requests like these?
> ref. [hashicorp/terraform#25133 (comment)](https://github.com/hashicorp/terraform/issues/25133#issuecomment-639131813)

Hmm, actually this is an interesting question. I will ask responsible people, and I will come back to you.

**Comment by toma-soft** [comment]
Regarding problem `#4` 

```
D:\some\path>terraform destroy -auto-approve -no-color 
data.docker_registry_image.jenkins-aec-image: Refreshing state... [id=sha256:da0371fa64e1c46ced0a98e648a8534791817d4271bc9a042fb5957a6ec69a11]
docker_volume.testjenkins: Refreshing state... [id=testjenkins]
docker_image.jenkins-aec: Refreshing state... [id=sha256:f6b4b2457a936ac66ced8159964676cfca52df6f1aa710c7cc3941e0d921f8c6some-registry/some-repository/jenkins:my-version]
docker_volume.testjenkins: Destroying... [id=testjenkins]
docker_image.jenkins-aec: Destroying... [id=sha256:f6b4b2457a936ac66ced8159964676cfca52df6f1aa710c7cc3941e0d921f8c6azdnuk-binss.dsuk.emea.denso/docker-cict-repository/jenkins-aec:2.277.1-release2]
docker_image.jenkins-aec: Destruction complete after 0s
docker_volume.testjenkins: Still destroying... [id=testjenkins, 10s elapsed]
docker_volume.testjenkins: Still destroying... [id=testjenkins, 20s elapsed]
Error: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
```

Unfortunately I forgot to enable debug logs, so I suspect you can't do much about it right now ? I will enable debug logs in all the pipelines.

**Comment by toma-soft** [comment]
I was "lucky" to get debug logs quickly regarding issue `#5`

```
2021-03-12T08:51:18.775+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:18 [INFO] Waiting for volume: 'testjenkins' to get removed: max '30 seconds'
2021-03-12T08:51:18.775+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:18 [DEBUG] Waiting for state to become: [removed]
docker_image.jenkins-aec: Destruction complete after 0s
2021-03-12T08:51:20.794+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:20 [INFO] Volume with id 'testjenkins' is still in use
2021-03-12T08:51:20.794+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:20 [TRACE] Waiting 5s before next try
2021-03-12T08:51:25.798+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:25 [INFO] Volume with id 'testjenkins' is still in use
2021-03-12T08:51:25.798+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:25 [TRACE] Waiting 10s before next try
docker_volume.testjenkins: Still destroying... [id=testjenkins, 10s elapsed]
2021-03-12T08:51:35.814+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:35 [INFO] Volume with id 'testjenkins' is still in use
2021-03-12T08:51:35.814+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:35 [TRACE] Waiting 10s before next try
docker_volume.testjenkins: Still destroying... [id=testjenkins, 20s elapsed]
2021-03-12T08:51:45.827+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:45 [INFO] Volume with id 'testjenkins' is still in use
2021-03-12T08:51:45.827+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:45 [TRACE] Waiting 10s before next try
2021-03-12T08:51:48.786+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:48 [WARN] WaitForState timeout after 30s
2021-03-12T08:51:48.786+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:48 [WARN] WaitForState starting 30s refresh grace period
docker_volume.testjenkins: Still destroying... [id=testjenkins, 30s elapsed]
2021/03/12 08:51:48 [DEBUG] docker_volume.testjenkins: apply errored, but we're indicating that via the Error pointer rather than returning it: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
2021/03/12 08:51:48 [ERROR] eval: *terraform.EvalApplyPost, err: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
2021/03/12 08:51:48 [ERROR] eval: *terraform.EvalSequence, err: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
2021/03/12 08:51:48 [ERROR] eval: *terraform.EvalOpFilter, err: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)

Error: timeout while waiting for state to become 'removed' (last state: 'in_use', timeout: 30s)
```

The volume couldn't be removed because the corresponding service was not removed earlier (also created by the same Terraform script).

```
2021/03/12 08:51:17 [INFO] terraform: building graph: GraphTypePlanDestroy
2021/03/12 08:51:17 [DEBUG] ProviderTransformer: "docker_image.jenkins" (*terraform.NodePlanDestroyableResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:17 [DEBUG] ProviderTransformer: "docker_volume.testjenkins" (*terraform.NodePlanDestroyableResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:17 [DEBUG] ProviderTransformer: "data.docker_registry_image.jenkins-image" (*terraform.NodePlanDestroyableResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:17 [DEBUG] Starting graph walk: walkPlanDestroy
2021-03-12T08:51:17.959+0100 [INFO]  plugin: configuring client automatic mTLS
2021-03-12T08:51:18.009+0100 [DEBUG] plugin: starting plugin: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe args=[.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe]
2021-03-12T08:51:18.119+0100 [DEBUG] plugin: plugin started: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe pid=24352
2021-03-12T08:51:18.119+0100 [DEBUG] plugin: waiting for RPC address: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe
2021-03-12T08:51:18.195+0100 [INFO]  plugin.terraform-provider-docker_v2.11.0.exe: configuring server automatic mTLS: timestamp=2021-03-12T08:51:18.191+0100
2021-03-12T08:51:18.224+0100 [DEBUG] plugin: using plugin: version=5
2021-03-12T08:51:18.224+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: plugin address: address=127.0.0.1:10000 network=tcp timestamp=2021-03-12T08:51:18.223+0100
2021-03-12T08:51:18.405+0100 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2021/03/12 08:51:18 [INFO] backend/local: apply calling Apply
2021/03/12 08:51:18 [INFO] terraform: building graph: GraphTypeApply
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "docker_volume.testjenkins (destroy)" (*terraform.NodeDestroyResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "docker_image.jenkins (destroy)" (*terraform.NodeDestroyResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "data.docker_registry_image.jenkins-image (destroy)" (*terraform.NodeDestroyResourceInstance) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "docker_image.jenkins (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "docker_volume.testjenkins (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "docker_service.testjenkins (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ProviderTransformer: "data.docker_registry_image.jenkins-image (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/12 08:51:18 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/kreuzwerker/docker\"]" references: []
2021/03/12 08:51:18 [DEBUG] ReferenceTransformer: "docker_image.jenkins (expand)" references: []
2021/03/12 08:51:18 [DEBUG] ReferenceTransformer: "docker_volume.testjenkins (expand)" references: []
2021/03/12 08:51:18 [DEBUG] ReferenceTransformer: "docker_service.testjenkins (expand)" references: []
2021/03/12 08:51:18 [DEBUG] ReferenceTransformer: "data.docker_registry_image.jenkins-image (expand)" references: []
2021/03/12 08:51:18 [DEBUG] pruneUnusedNodes: docker_image.jenkins (expand) is no longer needed, removing
2021/03/12 08:51:18 [DEBUG] pruneUnusedNodes: docker_volume.testjenkins (expand) is no longer needed, removing
2021/03/12 08:51:18 [DEBUG] pruneUnusedNodes: docker_service.testjenkins (expand) is no longer needed, removing
2021/03/12 08:51:18 [DEBUG] pruneUnusedNodes: data.docker_registry_image.jenkins-image (expand) is no longer needed, removing
2021/03/12 08:51:18 [DEBUG] Starting graph walk: walkDestroy
2021-03-12T08:51:18.422+0100 [INFO]  plugin: configuring client automatic mTLS
2021-03-12T08:51:18.460+0100 [DEBUG] plugin: starting plugin: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe args=[.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe]
2021-03-12T08:51:18.586+0100 [DEBUG] plugin: plugin started: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe pid=5304
2021-03-12T08:51:18.586+0100 [DEBUG] plugin: waiting for RPC address: path=.terraform/plugins/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe
2021-03-12T08:51:18.660+0100 [INFO]  plugin.terraform-provider-docker_v2.11.0.exe: configuring server automatic mTLS: timestamp=2021-03-12T08:51:18.656+0100
2021-03-12T08:51:18.689+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: plugin address: address=127.0.0.1:10001 network=tcp timestamp=2021-03-12T08:51:18.689+0100
2021-03-12T08:51:18.689+0100 [DEBUG] plugin: using plugin: version=5
2021-03-12T08:51:18.743+0100 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2021-03-12T08:51:18.773+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: 2021/03/12 08:51:18 [DEBUG] Using username for registry auths: ptm-ldap-mng01
2021/03/12 08:51:18 [DEBUG] EvalApply: ProviderMeta config value set
2021/03/12 08:51:18 [DEBUG] docker_volume.testjenkins: applying the planned Delete change
docker_volume.testjenkins: Destroying... [id=testjenkins]
```
I don't see in debug logs anything indicating the problem with removing docker_volume/docker_service.

**Comment by suzuki-shunsuke** [comment]
> The volume couldn't be removed because the corresponding service was not removed earlier (also created by the same Terraform script).

In the above Terraform Configuration https://github.com/kreuzwerker/terraform-provider-docker/issues/138#issue-829047064, there is no dependency between `docker_volume.prod-jenkins` and `docker_service.prod-jenkins`, so Terraform tries to remove them in parallel.

You should define the dependency  between `docker_volume.prod-jenkins` and `docker_service.prod-jenkins` explicitly.

```tf
resource "docker_service" "prod-jenkins" {
  name = "prod-jenkins"

  task_spec {
    container_spec {
      image = data.docker_registry_image.jenkins-image.name

      mounts {
        target    = "/var/jenkins_home"
        source    = docker_volume.prod-jenkins.name # docker_service.prod-jenkins depends on docker_volume.prod-jenkins
        type      = "volume"
        read_only = false
      }
      env = {
        # some env variables
      }
    }
  }
  # ...
}
```

Then docker_service.prod-jenkins will remove after docker_volume.prod-jenkins would be removed.

**Comment by toma-soft** [comment]
You have noticed the right thing, that the volume is being removed before the service.

But even after updating the script:

```
(...)
resource "docker_volume" "testjenkins" {
  name = "testjenkins"
}

resource "docker_service" "testjenkins" {
  name = "testjenkins"

  task_spec {
    container_spec {
      image = data.docker_registry_image.jenkins-image.name

      mounts {
        target    = "/var/jenkins_home"
        source    = docker_volume.testjenkins.name
        type      = "volume"
        read_only = false
      }

      env = {
         # some env variables
      }

    }

  }
  # ...
}
(...)
```
and in logs there the information which resources are referenced by the service:

`2021/03/12 09:50:14 [DEBUG] ReferenceTransformer: "docker_service.testjenkins (expand)" references: [data.docker_registry_image.jenkins-aec-image (expand) docker_volume.testjenkins (expand)]`

Still the volume is the first resource to be removed, not the service.

**Comment by suzuki-shunsuke** [comment]
https://github.com/kreuzwerker/terraform-provider-docker/issues/138#issuecomment-797329071

> Then docker_service.prod-jenkins will be removed **after** docker_volume.prod-jenkins would be removed.

Sorry, it's wrong.

docker_service.prod-jenkins depends on docker_volume.prod-jenkins,
so docker_service.prod-jenkins will be removed **before** docker_volume.prod-jenkins would be removed.

> Still the volume is the first resource to be removed, not the service.

It's strange...

ref. https://learn.hashicorp.com/tutorials/terraform/dependencies

**Comment by toma-soft** [comment]
I am trying to find the right dependentices, here is the information about references between resources:
```
2021/03/16 12:36:32 [INFO] terraform: building graph: GraphTypeValidate
2021/03/16 12:36:32 [DEBUG] ProviderTransformer: "docker_image.jenkins-aec" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/16 12:36:32 [DEBUG] ProviderTransformer: "docker_volume.testjenkins" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/16 12:36:32 [DEBUG] ProviderTransformer: "docker_service.testjenkins" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/16 12:36:32 [DEBUG] ProviderTransformer: "data.docker_registry_image.jenkins-aec-image" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/kreuzwerker/docker"]
2021/03/16 12:36:32 [DEBUG] ReferenceTransformer: "docker_service.testjenkins" references: [docker_volume.testjenkins data.docker_registry_image.jenkins-aec-image]
2021/03/16 12:36:32 [DEBUG] ReferenceTransformer: "data.docker_registry_image.jenkins-aec-image" references: []
2021/03/16 12:36:32 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/kreuzwerker/docker\"]" references: []
2021/03/16 12:36:32 [DEBUG] ReferenceTransformer: "docker_image.jenkins-aec" references: [data.docker_registry_image.jenkins-aec-image data.docker_registry_image.jenkins-aec-image]
2021/03/16 12:36:32 [DEBUG] ReferenceTransformer: "docker_volume.testjenkins" references: []
```

It is to see that docker_volume has no references while docker_service has many references. Hence docker_volume is tried be removed first. I will let you know when I set the right configuration.

**Comment by suzuki-shunsuke** [comment]
I outputted the dependency graph too.

```tf
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "2.11.0"
    }
  }
}

provider "docker" {
  host = "npipe:////.//pipe//docker_engine"
}

resource "docker_service" "prod-jenkins" {
  name = "prod-jenkins"

  task_spec {
    container_spec {
      mounts {
        target    = "/var/jenkins_home"
        source    = docker_volume.prod-jenkins.name # docker_service.prod-jenkins depends on docker_volume.prod-jenkins
        type      = "volume"
        read_only = false
      }
    }
  }
}

resource "docker_volume" "prod-jenkins" {
  name = "prod-jenkins"
}
```

```console
$ terraform version
Terraform v0.14.8
+ provider registry.terraform.io/kreuzwerker/docker v2.11.0

$ terraform graph | dot -Tsvg > graph.svg
```

![image](https://user-images.githubusercontent.com/13323303/111309575-678abe00-869f-11eb-838f-141586134cee.png)

The graph shows `docker_service.prod-jenkins` depends on `docker_volume.prod-jenkins` as we expect.

**Comment by toma-soft** [comment]
I have prepared the script similar to your version and so far I haven't seen the problem with removing the volume.
However, I have just noticed a new type of problem:

```
2021/03/16 22:01:50 [DEBUG] Starting graph walk: walkValidate
2021-03-16T22:01:50.467+0100 [INFO]  plugin: configuring client automatic mTLS
2021-03-16T22:01:50.591+0100 [DEBUG] plugin: starting plugin: path=.terraform/providers/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe args=[.terraform/providers/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe]
2021-03-16T22:01:50.694+0100 [DEBUG] plugin: plugin started: path=.terraform/providers/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe pid=6048
2021-03-16T22:01:50.694+0100 [DEBUG] plugin: waiting for RPC address: path=.terraform/providers/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe
2021-03-16T22:01:50.754+0100 [INFO]  plugin.terraform-provider-docker_v2.11.0.exe: configuring server automatic mTLS: timestamp=2021-03-16T22:01:50.751+0100
2021-03-16T22:01:50.897+0100 [DEBUG] plugin: using plugin: version=5
2021-03-16T22:01:50.897+0100 [DEBUG] plugin.terraform-provider-docker_v2.11.0.exe: plugin address: address=127.0.0.1:10000 network=tcp timestamp=2021-03-16T22:01:50.897+0100
2021-03-16T22:02:10.898+0100 [WARN]  plugin.stdio: stdio service not available, stdout/stderr syncing unavailable
2021-03-16T22:02:10.898+0100 [WARN]  plugin.stdio: stdio service unavailable, run will do nothing

Error: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: context deadline exceeded"


2021-03-16T22:02:12.905+0100 [WARN]  plugin: plugin failed to exit gracefully
2021-03-16T22:02:12.916+0100 [DEBUG] plugin: plugin process exited: path=.terraform/providers/registry.terraform.io/kreuzwerker/docker/2.11.0/windows_amd64/terraform-provider-docker_v2.11.0.exe pid=6048 error="exit status 1"
```
Do you know what causes that issue?

**Comment by suzuki-shunsuke** [comment]
> Do you know what causes that issue?

Sorry but I don't know.
Maybe this is related. https://github.com/kreuzwerker/terraform-provider-docker/issues/138#issuecomment-796754608

> Do you have any firewall software on your Windows system that might be blocking requests like these?
> ref. hashicorp/terraform#25133 (comment)

Could you confirm whether the same problem occurs with the other Terraform Provider like AWS Provider and MySQL Provider?
If the same problem occurs, this isn't a problem of this Docker Provider.

**Comment by toma-soft** [comment]
I can confirm that the graph helped me much. I could see how the items were related. Then I started to create the .tf file once again with full understanding the relations between volumes, services, etc.

**Comment by jackliu2006** [comment]
any solution to this issue already? I am having this issue also....
